### PR TITLE
OTA: chunked download

### PR DESCRIPTION
**Pull request by pennam**
This pull request introduces several changes to the OTA update process in the Arduino IoT Cloud library. The key changes focus on improving the reliability of OTA updates by adding support for chunked downloads and refining OTA policy management.

Improvements to OTA policy management:

* [`src/ArduinoIoTCloudTCP.h`](diffhunk://#diff-a1314de94eb0792cf8632064de47dae933ca9243976b03715c1083d1faba2389L99-R110): Replaced `setOtaPolicies` with `enableOtaPolicy` and `disableOtaPolicy` methods to manage OTA policies more flexibly. Added a new method `setOTAChunkMode` to enable or disable chunked downloads.
* [`src/ota/interface/OTAInterface.cpp`](diffhunk://#diff-6e3e6582c41e8451a16dae10757ac9278c9c91a65567aab491f8f0eee40f33c9L170-R173): Updated the `otaAvailable` method to use the new `getOtaPolicy` method for checking OTA policies.
* [`src/ota/interface/OTAInterface.h`](diffhunk://#diff-48a6e883d40d54b216295e81650bd93c9289d2671e2b5f0868eade4d0178c86cL83-R98): Added new methods `enableOtaPolicy`, `disableOtaPolicy`, and `getOtaPolicy` to manage OTA policies. Introduced a new `ChunkDownload` flag for chunked downloads.

Enhancements to OTA download process:

* [`src/ota/interface/OTAInterfaceDefault.cpp`](diffhunk://#diff-15f356c4407a55f1b72470304917cb77c9ade194e19c26e83731e93337047efcL44-R54): Refactored the `startOTA` and `fetch` methods to support chunked downloads. Added a new method `requestOta` to handle HTTP requests for OTA updates, including chunked requests. 
* [`src/ota/interface/OTAInterfaceDefault.h`](diffhunk://#diff-df4de9e6398b7f343066d7316cf57f994490a81c0b8e8b7adc02ff84d1d5d885L45-R47): Added new constants and methods to support chunked downloads, including `maxChunkSize`, `requestOta`, and `fetchMore`. Updated the `Context` struct to include new fields for managing chunked downloads.

### CODE REVIEW

File: src/ota/interface/OTAInterfaceDefault.cpp [hunk 1]
```diff
@@ -41,57 +42,43 @@ OTACloudProcessInterface::State OTADefaultCloudProcessInterface::startOTA() {
...
-  }
-
-  int statusCode = http_client->responseStatusCode();
-
-  if(statusCode != 200) {
-    DEBUG_VERBOSE("OTA ERROR: get response on \"%s\" returned status %d", OTACloudProcessInterface::context->url, statusCode);
-    return HttpResponseFail;
-  }
+  // make the http get request
+  requestOta(OtaFetchTime);
```
Comment (andreagilardoni) [inline src/ota/interface/OTAInterfaceDefault.cpp:53]: The return value is not being handled, an error could be returned

File: src/ota/interface/OTAInterfaceDefault.cpp [hunk 2]
```diff
@@ -153,13 +142,69 @@ OTACloudProcessInterface::State OTADefaultCloudProcessInterface::fetch() {
...
+  }
+
+  int statusCode = http_client->responseStatusCode();
+
+  if(((mode == OtaFetchChunk) && (statusCode != 206)) || ((mode == OtaFetchTime) && (statusCode != 200))) {
+    DEBUG_VERBOSE("OTA ERROR: get response on \"%s\" returned status %d", OTACloudProcessInterface::context->url, statusCode);
+    return HttpResponseFail;
+  }
+
+  http_client->skipResponseHeaders();
```
Comment (andreagilardoni) [inline src/ota/interface/OTAInterfaceDefault.cpp:188]: Doesn't this skip the the contentLength heade  that is being later used in line 56?

Comment (pennam) [reply src/ota/interface/OTAInterfaceDefault.cpp:188]: It just keep reading until the header is finished https://github.com/arduino-libraries/ArduinoHttpClient/blob/8b76ac2c31724a4f399576996cf5938f29991132/src/HttpClient.cpp#L510

contentLength() do exactly the same https://github.com/arduino-libraries/ArduinoHttpClient/blob/8b76ac2c31724a4f399576996cf5938f29991132/src/HttpClient.cpp#L546

**Comment by pennam** [comment]
@andreagilardoni would it be better to use `OtaFlags` instead of adding a new enum to select the download mode?

File: src/ota/interface/OTAInterfaceDefault.cpp (hunk 1 above)
Comment (pennam) [reply src/ota/interface/OTAInterfaceDefault.cpp]: fixed here https://github.com/arduino-libraries/ArduinoIoTCloud/pull/520/commits/89dddf7695485a766b54a4f481f35074e3411e37

**Comment by andreagilardoni** [comment]
> @andreagilardoni would it be better to use `OtaFlags` instead of adding a new enum to select the download mode?

I think you are right! this could be a good use for them,  since they are designed to accommodate settings for the OtaProcedure

**Review by andreagilardoni — APPROVED** [review]

**Pull request merged**